### ISSUE TITLE
Opt-in capability for my instance to notify me when it sees a post containing a link to one of my posts

### ISSUE THREAD

**Issue by afontenot** [mastodon/mastodon#20976]
### Pitch

The preferences page would contain a simple option with a check box: "Notify me when someone links to one of my posts."

When selected (off by default), the user would receive a notification from the instance if the instance sees a post (via local post or federation) that links to one of the user's posts. (Due to federation not every post on the Fediverse would trigger such a notification. This is an understood limitation and not part of the feature request.)

### Motivation

One reason many users don't like the idea of quote boosting is because of the threat of harassment. However, there are two approaches different users may take towards harassment:

 * They may wish never to know if someone is linking to their posts in a mocking way or trying to get others to go attack them. In this case, the status quo is fine.
 * They may wish to be actively informed that this is happening, so that they can take steps to prevent it (e.g. by blocking the user and reporting them to the moderators). In this latter case, Mastodon doesn't currently do much to help.

Supporters of the latter are among those requesting the implementation of quote boosting on Mastodon in issues like https://github.com/mastodon/mastodon/issues/20673. I think their position is reasonable, but I think this kind of anti-harassment capability could be built (via this feature request) in a way that doesn't require (but is compatible with) the controversial quote boosting feature to be implemented. In particular:

 * I'm not asking for a feature that would allow certain posts to be displayed in line with a "quoting" post. Linked posts would remain the same under this feature request.
 * I'm not asking for a feature that would make it easier to link to another user's posts, by providing a button designed for doing so on the post.

This request would also benefit users who may want to engage (positively) with others who link to their posts. This would enable some of that without making quote boosting an inescapable part of the Mastodon / Fediverse experience.

**Comment by afontenot** [comment]
Something worth mentioning here since a lot of the debate around this feature is happening on the "quote-boosting" suggestion: some users have argued that there are sometimes good reasons to not want a user to be notified when you're linking to them. I think there are three questions that need to be answered under the general heading of this feature request:

1. Do we want users to have the option to be notified when someone links to them (when possible because the instance is aware of such a link):
2. Do we want to make this option opt-in / opt-out for the users in question? How does the behavior interact with a block of the linking user?
3. Do we want it to be possible for a user to link to an opted-in user *without* sending them a notification in an otherwise-public post, or should the notification be completely within the control of the linked-to user?

(3) strikes me as particularly fraught. On the one hand it might be useful to link to a network of accounts participating in harassment. On the other hand, the assumption (hope?) that a public post will not be noticed by someone seems like a bad one that should be discouraged - after all, someone might have an account following the public timeline with an app set up specifically to extract links to their content. There's no way to prevent this.

**Comment by trwnh** [comment]
I can see an issue with people copy-pasting the proxy links which would not be detected unless redirects were taken into account somehow. But I'll also mention that there was a proposal to use WebMentions at some point in #6074 which ended up getting rejected: https://github.com/mastodon/mastodon/issues/6074#issuecomment-378452136

> I did a poll for public opinion: https://mastodon.social/@Gargron/99798025215868039 And while the actual poll is almost 50:50, a lot of valid privacy concerns were raised
> - It's a privacy violation that people aren't expecting (change of behaviour, not desired behaviour for everybody)
> - It's a way to potentially DDoS websites in coordinated attacks
> - It has processing costs (HTTP requests are always nasty because they can take multiple seconds or even longer if the website is unresponsive)
> - Hiding it behind options would mean it's useless to website authors as barely anyone would discover that option or care to enable it. For that matter, it's a waste of development time for Mastodon too if almost nobody will use it

On the receiving end there's #1384 which also went nowhere.

Aside from WebMentions, Mastodon could try to resolve/fetch/lookup every single link as an ActivityPub object, and if an object link is discovered, then it could get exposed via the API as a tag similar to mentions and hashtags. But sending a notification for such would carry the same privacy implications as above -- people would suddenly have to be careful who they merely *link* to, even without mentioning or addressing them. In that regard, the Webmention approach is a bit better in that it requires both the linker to opt-in to sending Webmentions, and the linkee to opt-in to receiving Webmentions.

**Comment by afontenot** [comment]
> Aside from WebMentions, Mastodon could try to resolve/fetch/lookup every single link as an ActivityPub object, and if an object link is discovered

Yeah, fair point. There will always be ways to hide a link back to a user's profile or posts. To some extent this option has to assume good faith, which limits its potential as an anti-harassment feature.

**Comment by onpon4** [comment]
This proposal is essentially to trick people into @ mentioning others, and I think is a wrong approach that is more likely to aid abusers than it is to hinder harassers.

Consider: if someone regularly organizes harassment by linking to people, and they notice that the people they're trying to organize harassment against is able to instantly block them because the link notifies their target, all they would have to do is obscure the link so that it doesn't trigger the notification. What effort they would have to go to would depend on how far the implicit notification system goes to stop these efforts, but there will always be a way to thwart it fairly easily (servers running Mastodon don't have infinite computing resources after all).

But then, consider an example of an abuse victim coming out about an experience with a particularly influential abuser. They post a link to something the abuser posted, not wishing to notify them. And let's say the victim is on an instance type that isn't Mastodon, so Mastodon can't warn them that the link implies what is effectively an @ mention. Now the abuser knows immediately about the call-out and starts spreading a rumor about the victim to preemptively discredit them.

It seems to me that the way for Mastodon to cause a "quote-boosting" button (were it implemented) to notify those being "quote-boosted" would be very simple: just include an @ mention of the "quote-boosted" post's author in the message entry textbox by default, exactly the same as replies. Because you can't force someone to write their post in a way that notifies the original author anyway, and it's not always in the social good to notify the original author. That's a nuanced question that moderators have to handle.

🦇

**Comment by afontenot** [comment]
> This proposal is essentially to trick people into @ mentioning others, and I think is a wrong approach that is more likely to aid abusers than it is to hinder harassers.

I think I understand your argument; I'm not sure I agree, but I get your concerns. But I don't understand why this is a "trick". It's just a matter of establishing expectations on social media. Let me explain why.

> But then, consider an example of an abuse victim coming out about an experience with a particularly influential abuser. They post a link to something the abuser posted, not wishing to notify them.

Well, suppose this user instead said "`@username` has been harassing me". And suppose this user didn't know that posting an "@" did more than automatically link to the profile. They might feel that they had been tricked into notifying their abuser abut their post. It might seem like Mastodon was out to get them or was aiding abusers. But of course this wouldn't be the case. Their expectations for how the platform operates were just ... incorrect. I think in this case we would say that this result is unfortunate, but we wouldn't say Mastodon was to blame for what happened.

Currently, the user expectation that linking to someone's posts will *not* trigger a notification is basically unfounded. It's *unlikely* that it will, but ultimately a harasser or one of their allies *could* be monitoring select hashtags (e.g. `#fediblock`) or the entire public timeline, or they could be running their own instance, or what have you. In any of these cases implementing "send me a notification whenever anyone links to me" would be absolutely trivial. 

On the other hand, I made this issue because a lot of people said they needed this feature to help combat harassment they were facing. If this feature was implemented in the way you describe but made optional, it wouldn't serve that purpose, since it could trivially be disabled by removing the "@". I think the principle is "if someone on the Fediverse is sending a whole bunch of unwanted attention my way by linking directly to me, I should know about it".

Again, I'm not 100% convinced this feature is a great idea, and I think you have some good arguments, but that's the claim.

**Comment by onpon4** [comment]
The difference with @ mentions is that @ mentions are well established throughout the fediverse to be something that causes a notification. That hasn't been established for merely posting links. It *could* be firmly established, and then it wouldn't really be a problem, but consider: where does that leave us?

Those who don't want to notify the person they're talking about will simply mangle the links so that they don't find out anyway.

Long-term, the *only* true benefit of making links to people's post behave like @ mentions is making posts where people today would include an explicit @ mention slightly shorter. That's only *after* the entire fediverse decides that "links are @ mentions" is the way to go (assuming such a consensus is even possible). And that comes at a cost of links being broken up or otherwise mangled to subvert the implicit @ mention, *sometimes for a legitimate reason*, thus making the experience less accessible for no good reason.

In the long-term, it is always going to require a poster's consent to notify anyone else, even someone they're talking about, and it always should. So I don't see a good reason to go through all the effort of shifting the entire fediverse to make trawling through posts searching for URLs the norm, when the established system of @ mentioning can be used instead much more easily.

🦇

**Comment by afontenot** [comment]
@onpon4 I think that's a good summary of the "oppose" argument, thanks!
